File: app/src/main/java/io/github/lmller/codetalks2017/MainActivity.java
package io.github.lmller.codetalks2017;

import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.widget.SearchView;

import javax.inject.Inject;

import io.github.lmller.codetalks2017.databinding.ActivityMainBinding;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;
import io.reactivex.subjects.PublishSubject;

public class MainActivity extends AppCompatActivity {

	@Inject
	OpenThesaurus thesaurus;

	private PublishSubject<String> searchQueryChangedSubject = PublishSubject.create();
	private SearchViewModel viewModel;

	private Disposable disposable;

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		App.getInjector().injectInto(this);

		ActivityMainBinding binding = DataBindingUtil.setContentView(this, R.layout.activity_main);
		viewModel = new SearchViewModel(thesaurus);

		binding.setViewModel(viewModel);

		binding.search.setOnQueryTextListener(new SearchView.OnQueryTextListener() {

			@Override
			public boolean onQueryTextSubmit(String query) {
				return true;
			}

			@Override
			public boolean onQueryTextChange(String text) {
				searchQueryChangedSubject.onNext(text.trim());
				return true;
			}
		});

		disposable = viewModel.observeTextChanges(searchQueryChangedSubject, Schedulers.io());
	}

	@Override
	protected void onDestroy() {
		super.onDestroy();
		disposable.dispose();
	}
}

File: app/src/main/java/io/github/lmller/codetalks2017/MainActivity.kt
package io.github.lmller.codetalks2017

import android.databinding.DataBindingUtil
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.widget.SearchView
import io.github.lmller.codetalks2017.databinding.ActivityMainBinding
import io.reactivex.disposables.Disposables
import io.reactivex.schedulers.Schedulers
import io.reactivex.subjects.PublishSubject
import javax.inject.Inject

class MainActivity : AppCompatActivity() {

    @Inject
    lateinit var thesaurus: OpenThesaurus

    private val searchQueryChangedSubject = PublishSubject.create<String>()
    private var searchViewModel: SearchViewModel? = null
    private var disposable = Disposables.disposed()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        App.injector.injectInto(this)

        searchViewModel = SearchViewModel(thesaurus).apply {
            disposable = observeTextChanges(searchQueryChangedSubject, Schedulers.io())
        }

        with(DataBindingUtil.setContentView<ActivityMainBinding>(this, R.layout.activity_main)) {
            viewModel = searchViewModel

            search.setOnQueryTextListener(object : SearchView.OnQueryTextListener {

                override fun onQueryTextSubmit(query: String) = true

                override fun onQueryTextChange(text: String): Boolean {
                    searchQueryChangedSubject.onNext(text.trim())
                    return true
                }
            })
        }

    }

    override fun onDestroy() {
        super.onDestroy()
        disposable.dispose()
    }
}
